### PR TITLE
Break apart the `ComputeResource` entity class

### LINKED ISSUE #76: Create a new LibvirtComputeResource entity
I believe it is time to revisit the idea of abstracting ``entities.ComputeResource`` so that we can create specific entities to handle different types of compute resources. To get things started I'd like to request the creation of a ``entities.LibvirtComputeResource`` that can be used as such:

```python
ComputeResource(
    name=u'local_libvirt',
    location=[loc],
    organization=[org],
    provider=u'Libvirt',
    set_console_password=True,
    url=u'qemu+tcp://localhost:16509/system',
).create()
```
The ``GET /api/v2/compute_resources/:id`` call returns:

```python
{u'compute_attributes': [],
 u'created_at': u'2015-05-15T21:01:46Z',
 u'description': u'',
 u'display_type': u'VNC',
 u'id': 1,
 u'images': [],
 u'locations': [{u'id': 2,
   u'name': u'Default Location',
   u'title': u'Default Location'}],
 u'name': u'local_libvirt',
 u'organizations': [{u'id': 1,
   u'name': u'Default Organization',
   u'title': u'Default Organization'}],
 u'provider': u'Libvirt',
 u'provider_friendly_name': u'Libvirt',
 u'set_console_password': True,
 u'updated_at': u'2015-05-15T21:25:17Z',
 u'url': u'qemu+tcp://localhost:16509/system'}
```

### PR DESCRIPTION
**Pull request by Ichimonji10**
Fix #76:

> Create a new LibvirtComputeResource entity
>
> I believe it is time to revisit the idea of abstracting
> `entities.ComputeResource` so that we can create specific entities to handle
> different types of compute resources. To get things started I'd like to
> request the creation of a `entities.LibvirtComputeResource` […]

Remove the `ComputeResource` entity class and create the following three classes
in its stead:

* `AbstractComputeResource`
* `DockerComputeResource`
* `LibvirtComputeResource`

Example of usage:

```python
>>> from nailgun import entities
>>> dcr = entities.DockerComputeResource().create(True)
>>> lcr = entities.LibvirtComputeResource().create(True)
>>> dcr.read()
nailgun.entities.DockerComputeResource(
    nailgun.config.ServerConfig(
        auth=('admin', 'changeme'),
        url='https://…',
        verify=False,
        version=<Version('6.1.0')>,
    ),
    description=None,
    email=None,
    id=34,
    location=[],
    name='芧顲鴧匕冻决翌',
    organization=[],
    provider='Docker',
    provider_friendly_name='Docker',
    url='https://test.com',
    user=None,
)
>>> lcr.read()
nailgun.entities.LibvirtComputeResource(
    nailgun.config.ServerConfig(
        auth=('admin', 'changeme'),
        url='…',
        verify=False,
        version=<Version('6.1.0')>,
    ),
    description=None,
    display_type='SPICE',
    id=35,
    location=[],
    name='掄簴蟉鈂鞾苠飐枾叛谒鲜嬪砛还燳琻醾銚忽',
    organization=[],
    provider='Libvirt',
    provider_friendly_name='Libvirt',
    set_console_password=True,
    url='ftp://example.info',
)
```

### CODE REVIEW

File: nailgun/entities.py [hunk 1]
```diff
@@ -423,74 +422,96 @@ def __init__(self, server_config=None, **kwargs):
...
+        For more information, see `Bugzilla #1223540
+        <https://bugzilla.redhat.com/show_bug.cgi?id=1223540>`_.
+
+        Also, do not try to read the "password" field. No value is returned for
+        the field, for obvious reasons.
+
+        """
+        if attrs is None:
+            attrs = self.read_json()
+        if 'email' not in attrs and 'email' not in ignore:
```
Comment (omaciel) [inline nailgun/entities.py]: Why do we do a `PUT` here? This section of the code is not very clear to me :/

**Comment by elyezer** [comment]
What will happen if I use `entities.LibvirtComputeResource(id=<docker_compute_resource_id>).read()`?

File: nailgun/entities.py [hunk 2]
```diff
@@ -423,74 +422,96 @@ def __init__(self, server_config=None, **kwargs):
...
-            #
-            # 1. File a bugzilla issue asking for a fix.
-            # 2. Figure out what data is necessary and add it here.
-            pass
-        elif self.provider == 'Vmware':
-            for field in ('password', 'user', 'uuid'):
-                if not hasattr(self, field):
-                    setattr(self, field, self._fields[field].gen_value())
+
+class LibvirtComputeResource(AbstractComputeResource):
```
Comment (omaciel) [inline nailgun/entities.py]: Do we need to override the `read` method here too? IIRC there are some extra attributes that get returned by the server that don't seem to be handled here, such as `compute_attributes` and `images`.

**Comment by Ichimonji10** [comment]
> What will happen if I use `entities.LibvirtComputeResource(id=<docker_compute_resource_id>).read()`?

You'll get a traceback.

NailGun can gracefully deal with extra attributes being returned, but it blows up if not enough attributes are returned. Docker compute resources do not return all of the attributes needed to populate a libvirt compute resource and vice versa. If you want to read some compute resource but are not sure whether it's a docker compute resource, libvirt compute resource, etc, then use `AbstractComputeResource`. I did that during manual testing:

```python
>>> entities.AbstractComputeResource(id=1).read().provider
'Libvirt'
>>> entities.AbstractComputeResource(id=2).read().provider
'Docker'
```

File: nailgun/entities.py (hunk 1 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: When an HTTP GET request is issued to `/api/v2/compute_resources` for a docker compute resource, an "email" attribute should be returned. None is returned. However, the "email" attribute is returned in response to `PUT` requests. So we make `PUT` request here, grab the email, and insert the email into `attrs`.

File: nailgun/entities.py (hunk 2 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: NailGun ignores any extra attributes that are returned. In addition to `compute_attributes` and `images`, the server also returns the `created_at` and `updated_at` attributes, and we ignore them too.

File: nailgun/entities.py (hunk 1 above)
Comment (omaciel) [inline nailgun/entities.py]: Wow! Safe to assume that you'll file an issue with TheForeman?

Comment (Ichimonji10) [inline nailgun/entities.py]: I filed [BZ 1223540](https://bugzilla.redhat.com/show_bug.cgi?id=1223540). Would you like me to file a matching issue with TheForeman?

Comment (omaciel) [inline nailgun/entities.py]: No, this is perfect :)

**Comment by omaciel** [comment]
APT

File: nailgun/entities.py (hunk 1 above)
Comment (Ichimonji10) [inline nailgun/entities.py]: :+1:

**Comment by elyezer** [comment]
ACK

**Pull request merged**